lib/ProductOpener/APIProductRevert.pm

```diff
@@ -0,0 +1,181 @@
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::APIProductRevert - implementation of API to revert a product to a specific revision
+=head1 DESCRIPTION
+
+=cut
+
+package ProductOpener::APIProductRevert;
+
+use ProductOpener::PerlStandards;
+use Exporter qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN {
+	use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+		&revert_product_api
+	);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Display qw/:all/;
+use ProductOpener::Users qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Products qw/:all/;
+use ProductOpener::API qw/:all/;
+use ProductOpener::Text qw/:all/;
+use ProductOpener::Tags qw/:all/;
+use ProductOpener::Mail qw/:all/;
+
+use Encode;
+
+=head2 revert_product_api()
+
+Process API v3 requests to revert a product to a specific revision.
+
+=cut
+
+sub revert_product_api ($request_ref) {
+
+	$log->debug("revert_product_api - start", {request => $request_ref}) if $log->is_debug();
+
+	my $response_ref = $request_ref->{api_response};
+
+	# Set the result field for the API response, will be updated later if the product is successfully reverted
+	$response_ref->{result} = {id => "product_not_reverted"};
+
+	my $error = 0;
+
+	my $request_body_ref = $request_ref->{body_json};
+
+	# Check we have all needed input parameters
+	foreach my $field ("code", "rev") {
+		if (not defined $request_body_ref->{$field}) {
+			$log->error("revert_product_api - missing input $field", {request => $request_ref})
+				if $log->is_error();
+			add_error(
+				$response_ref,
+				{
+					message => {id => "missing_field"},
+					field => {id => $field},
+					impact => {id => "failure"},
+				}
+			);
+			$error = 1;
+		}
+	}
+
+	# Check that the user has permission (is an admin or a moderator, or we are on the producers platform)
+
+	$error += check_user_permission($request_ref, $response_ref, "product_revert");
+
+	if (not $error) {
+
+		# Load the product
+		my ($code, $ai_data_string) = &normalize_requested_code($request_body_ref->{code}, $response_ref);
+		my $rev = $request_body_ref->{rev};
+
+		# Check if the code is valid
+		if ($code !~ /^\d{4,24}$/) {
```

Comment:
We really have to centralize this kind of checks in Product.pm as we may extend codes in a few months !
So make it a method !

Reply:
Good point.